lib/blockchainmonitor.js

```diff
@@ -5,100 +5,107 @@ var _ = require('lodash');
-    opts[network] = opts[network] || {};
-    self.subscriber[network] = self._getAddressSubscriber(
-      opts[network].provider, network, opts[network].url);
-  });
-};
 
-nodeutil.inherits(BlockchainMonitor, events.EventEmitter);
+  async.parallel([
+
+    function(done) {
+      self.explorers = _.map(['livenet', 'testnet'], function(network) {
+        var config = opts.blockchainExplorerOpts[network] || {};
+        return self._initExplorer(config.provider, network, config.url);
+      });
+      done();
+    },
+    function(done) {
+      self.storage = new Storage();
+      self.storage.connect(opts.storageOpts, done);
+    },
+    function(done) {
+      self.messageBroker = new MessageBroker(opts.messageBrokerOpts);
+      done();
+    },
+  ], cb);
 
-BlockchainMonitor.prototype._getAddressSubscriber = function(provider, network) {
+};
+
+BlockchainMonitor.prototype._initExplorer = function(provider, network, url) {
   $.checkArgument(provider == 'insight', 'Blockchain monitor ' + provider + ' not supported');
 
+  var self = this;
+
   var explorer = new BlockchainExplorer({
     provider: provider,
     network: network,
+    url: url,
   });
 
   var socket = explorer.initSocket();
+  socket.emit('subscribe', 'inv');
+  socket.on('tx', _.bind(self._handleIncommingTx, self));
 
-  // TODO: Extract on its own class once more providers are implemented
-  return {
-    subscribe: function(address, handler) {
-      socket.emit('subscribe', address);
-      socket.on(address, handler);
-    },
-  };
+  return explorer;
 };
 
-BlockchainMonitor.prototype.subscribeAddresses = function(walletId, addresses) {
-  $.checkArgument(walletId);
-
+BlockchainMonitor.prototype._handleIncommingTx = function(data) {
   var self = this;
 
-  if (!addresses || addresses.length == 0) return;
+  if (!data || !data.vout) return;
 
-  function handlerFor(address, txid) {
-    var notification = Notification.create({
-      walletId: this,
-      type: 'NewIncomingTx',
-      data: {
-        address: address,
-        txid: txid,
-      },
-    });
-    self.emit('notification', notification);
-  };
+  var outs = _.compact(_.map(data.vout, function(v) {
+    var addr = _.keys(v)[0];
+    if (addr.indexOf('3') != 0 && addr.indexOf('2') != 0) return;
 
-  if (!self.subscriptions[walletId]) {
-    self.subscriptions[walletId] = {
-      addresses: [],
+    return {
+      address: addr,
+      amount: +v[addr]
     };
-  };
-
-  var addresses = [].concat(addresses);
-  var network = Bitcore.Address.fromString(addresses[0]).network.name;
-  var subscriber = self.subscriber[network];
-  _.each(addresses, function(address) {
-    self.subscriptions[walletId].addresses.push(address);
-    subscriber.subscribe(address, _.bind(handlerFor, walletId, address));
+  }));
+  if (_.isEmpty(outs)) return;
+
+  async.each(outs, function(out, next) {
+    self.storage.fetchAddress(out.address, function(err, address) {
+      if (err || !address) return next(err);
```

Comment:
if err we should die; or if the mongo driver try to reconnect itself, we should log.